reference-delegate-explainer.md

```diff
@@ -0,0 +1,534 @@
+# Reference Delegate for Cross-root ARIA
+
+Author: [Ben Howell](https://github.com/behowell)
+
+## Introduction
+
+Reference Delegate is a new HTML feature to help solve the cross-root ARIA problem. For background on cross-root ARIA, see @alice's article [How Shadow DOM and accessibility are in conflict](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/). The article describes the two main problems that need to be solved: [Referring from Shadow DOM outwards](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/#referring-from-shadow-dom-outwards) and [Referring into Shadow DOM](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/#referring-into-shadow-dom).
+
+The Reference Delegate proposal is based heavily on @Westbrook's [Cross-root ARIA Reflection API](https://github.com/Westbrook/cross-root-aria-reflection/blob/main/cross-root-aria-reflection.md) propsal, as well as borrowing ideas from @alice's [Semantic Delegate](https://github.com/alice/aom/blob/gh-pages/semantic-delegate.md) proposal.
+
+A complete solution to cross-root ARIA combines the existing [**ARIAMixin IDL attributes**](https://w3c.github.io/aria/#ARIAMixin) (such as `ariaLabelledbyElements` and `ariaActiveDescendantElement`) to solve the "Referring from Shadow DOM outwards" problem, and introduce a new feature **Reference Delegate** to solve the "Referring into Shadow DOM" problem, which is compatible with the ARIA.
+
+## Proposal
+
+### Reference Delegate
+
+Reference Delegate is a new feature that enables creating ARIA links to elements inside a component's shadow DOM, while maintaining encapsulation of the internal details of the shadow DOM.
+
+**Goals**
+
+- Create a mechanism for ID reference attributes like `aria-activedescendant` and `for` to refer to an element inside a component's shadow DOM.
```

Comment:
Is there a path by which the `referenceDelegate` could receive other of the attributes from the host that are required to achieve accessible patterns (e.g. `role` or `aria-label`, etc.) without them being ID references?

Reply:
@alice re: https://github.com/WICG/aom/pull/207#discussion_r1425951582, yes. I ask it more clearly here, as I found the full combobox example at the end and deleted the other comment, sorry!

I realize that the explainer says that it _doesn't_ do this, but it seems like a thing that people would want it to do.

Reply:
I specifically left that out of scope of this proposal. This proposal only solves the problem of IDREF attributes that _target_ the host element, as that is the only "missing piece" blocking use of ARIA with the shadow DOM. It does not address attributes that are _on_ the host element itself.

It's possible to implement something similar to what you mention, today using custom observed attributes, and passing them to elements inside the shadow tree. Notably, the ARIAMixin attributes like `ariaActiveDescendantElement`, combined with custom attributes, allow you to refer OUT of the shadow DOM. That said, the current state isn't perfect: It doesn't let you use the "real" attributes like `aria-label` or `role`. Instead, you need to invent custom attributes to forward to the elements inside. But it is at least doable today.

In an earlier draft of this proposal, I had an appendix with a possible declarative syntax to achieve the same result. I removed it from this proposal in order to avoid distracting the main proposal. I've moved it to a separate file here: [Declarative syntax for referring OUT of the Shadow DOM](https://github.com/behowell/aom/blob/attribute-binding/attribute-binding-explainer.md). However, that is still very early and still needs a lot of work.